kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -199,6 +210,10 @@ private Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata(
                             return new Tuple2<>(protocol, metadata);
                         }
                     } else {
+                        if (columnarBatch == null) {
+                            columnarBatch = nextElem.getFileDataReadResult().getData();
+                            assert(columnarBatch.getSchema().equals(PROTOCOL_METADATA_READ_SCHEMA));
```

Comment:
Why do we need to assert the schema? Is there some scenario where it would not match?

Reply:
This is part of a bigger topic. We have this opage `ColumnarBatch columnarBatch` used all throughout kernel.

Say I have a method, `foo`, that takes as input some columnarBatch. What is it's schem? As a developer, how do I write code that uses `foo`? I need to know what columns it has. How do we check this? How do we define this? 

`assert` is one way to do it. do you prefer something else? or nothing at all?

Reply:
Yeah, you're right that assert makes sense. It's not a user error when this happens, but rather a connector bug. In _theory_ the connector should have honored kernel's request for a read with the given schema, but in practice a connector bug would lead to arbitrary "fun" if we didn't notice early.